src/tls_stream.rs

```diff
@@ -588,12 +622,71 @@ impl<S> TlsStream<S>
                     self.out_buf.set_position(0);
                     Ok(())
                 }
-                err => Err(io::Error::from_raw_os_error(err as i32)),
+                err => {
+                    let err = io::Error::from_raw_os_error(err as i32);
+                    debug!("failed to encrypt: {:?}", err);
+                    Err(err)
+                }
             }
         }
     }
 }
 
+impl<S> MidHandshakeTlsStream<S>
+    where S: Read + Write,
+{
+    /// Returns a shared reference to the inner stream.
+    pub fn get_ref(&self) -> &S {
+        self.inner.get_ref()
+    }
+
+    /// Returns a mutable reference to the inner stream.
+    pub fn get_mut(&mut self) -> &mut S {
+        self.inner.get_mut()
+    }
+
+    /// Restarts the handshake process.
+    pub fn handshake(mut self) -> Result<TlsStream<S>, HandshakeError<S>> {
+        match self.inner.initialize() {
+            Ok(_) => Ok(self.inner),
+            Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
+                debug!("handshake incomplete");
+                Err(HandshakeError::Interrupted(self))
+            }
+            Err(e) => {
+                debug!("handshake error: {:?}", e);
+                Err(HandshakeError::Failure(e))
+            }
+        }
+    }
+}
+
+impl<S: fmt::Debug + Any> Error for HandshakeError<S> {
+    fn description(&self) -> &str {
+        match *self {
+            HandshakeError::Failure(_) => "failed to perform handshake",
+            HandshakeError::Interrupted(_) => "interrupted performing handshake",
+        }
+    }
+
+    fn cause(&self) -> Option<&Error> {
+        match *self {
+            HandshakeError::Failure(ref e) => Some(e),
+            HandshakeError::Interrupted(_) => None,
+        }
+    }
+}
+
+impl<S> fmt::Display for HandshakeError<S> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let msg = match *self {
```

Comment:
I have never seen anything print the entire error chain :P

This is a problem in itself, but in the meantime the convention seems to be something like this:

```rust
fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
    try!(fmt.write_str(self.description));
    if let Some(e) = self.cause() {
       try!(write!(fmt, ": {}", e));
    }
    Ok(())
}
```